File: cpp/0129_sumNumbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// 递归：每次修改根节点的值为父节点的值*10+根节点的值，最后对叶子节点求和
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int res = 0;
        TreeNode* parent = nullptr;
        traverse(root,parent,res);
        return res;
    }
    void traverse(TreeNode* root, TreeNode* parent,int &res)
    {
        if(root==nullptr)
            return;
        if(parent!=nullptr)
            root->val += parent->val * 10;
        if(root->right==nullptr && root->left==nullptr)
            res += root->val;
        traverse(root->left, root, res);
        traverse(root->right, root, res);
    }
};

File: java/0129_sumNumbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        int res = traverse(root,new TreeNode());
        return res;
    }
    public int traverse(TreeNode root,TreeNode parent)
    {
        if(root==null)
            return 0;
        root.val += parent.val*10;
        if(root.left==null && root.right==null)
            return root.val;
        else
            return traverse(root.left,root)+traverse(root.right,root);
    }
}
